docs/reference/ccr/apis/follow/put-follow.asciidoc

```diff
@@ -76,6 +76,26 @@ referenced leader index. When this API returns, the follower index exists, and
   (Required, string) The <<remote-clusters,remote cluster>> containing
   the leader index.
 
+[[ccr-put-follow-request-body-data_stream_name]]`data_stream_name`::
```

Comment:
Hm, if I understand correctly, CCR works on index level. With this API what we do is to follow a single leader index that belongs to a data stream and we provide also a data stream name. I do not see from this doc, any guidance on naming the follower index or the data stream of the follower index so it could be anything right?

Furthermore, I was wondering, if CCR is defined on an index level, this means that after the leader index will be rolled over the leader data stream would not be followed anymore. Right?

So, I was wondering. Is this API maybe too "free" and error prone, users might put invalid names and potentially create weird bugs? 

Should we instead, verify the name of the follower index that it follows the data stream pattern and extract the data stream name from it?

So, if a user tries to follow an data stream backing index, for example, `.ds-logs-mysql-default_copy-2022-01-01-000001` we will infer `logs-mysql-default_copy` as the data stream name. While in the case they try to create a follower index like `.ds-logs-mysql-default-2022-01-01-000001_copy`, we will fail and let them now that this is an invalid name.

Reply:
> Should we instead, verify the name of the follower index that it follows the data stream pattern and extract the data stream name from it?

It's worse than that, a data stream can have *any* arbitrary backing index name, because we have the [migrate to data stream](https://www.elastic.co/guide/en/elasticsearch/reference/8.3/indices-migrate-to-data-stream.html) API that converts an alias to a data stream. So a data stream named `logs-mysql-info` could have backing indices named `myfirstindex` and `other-index` following no pattern. Rather than enforce that there is a pattern with the backing index names, I opted to go with a check for something that *looks* like our regular pattern and if that doesn't work, then we treat it as an "I don't know" pattern and just do the regular name replacement.

If for some reason we have something that *looks* like a good pattern name, but is not, then we throw an exception and stop the auto-following. I'll add a test for this specific scenario (in e3001d50afaa2537ad51d58cbd9761811ffcd4f7).

> Furthermore, I was wondering, if CCR is defined on an index level, this means that after the leader index will be rolled over the leader data stream would not be followed anymore. Right?

If the leader is rolled over then the auto-follow pattern would pick up the next index in the data stream (presumably `<thing>-000002`), the data stream is not really "followed" because CCR doesn't really follow data streams, but the next index would be replicated, only the data stream would already exist because the first index would have created it already.

Things are a little different in ILM, because ILM auto-injects some `unfollow` actions around rollover, but regardless the naming thing would be the same.